File: Arrays/849. Maximize Distance to Closest Person.go
package main

import "math"

func Min(x, y int) int {
	if x > y {
		return y
	}
	return x
}

func Max(x, y int) int {
	if x < y {
		return y
	}
	return x
}

func maxDistToClosest(seats []int) int {
	n := len(seats)

	left, right := make([]int, n), make([]int, n)

	for i, _ := range left {
		left[i], right[i] = math.MaxInt32, math.MaxInt32
	}

	for i := 0; i < n; i++ {
		if seats[i] == 1 {
			left[i] = 0
		} else {
			previous := Min(0, i-1)
			if left[previous] != math.MaxInt32 {
				left[i] = left[previous] + 1
			}
		}
	}

	for i := n - 1; i >= 0; i-- {
		if seats[i] == 1 {
			right[i] = 0
		} else {
			next := Min(n-1, i+1)
			if right[next] != math.MaxInt32 {
				right[i] = right[next] + 1
			}
		}
	}

	distance := 0
	for i, _ := range left {
		if seats[i] != 1 {
			// fmt.Println(distance, left[i], right[i], i)
			distance = Max(distance, Min(left[i], right[i]))
		}
	}
	return distance
}

File: Arrays/849. Maximize Distance to Closest Person.cpp
/*
 ____  _     _           _     _       __  __       _ _           _
|  _ \(_)___| |__   __ _| |__ | |__   |  \/  | __ _| | |__   ___ | |_ _ __ __ _
| |_) | / __| '_ \ / _` | '_ \| '_ \  | |\/| |/ _` | | '_ \ / _ \| __| '__/ _` |
|  _ <| \__ \ | | | (_| | |_) | | | | | |  | | (_| | | | | | (_) | |_| | | (_| |
|_| \_\_|___/_| |_|\__,_|_.__/|_| |_| |_|  |_|\__,_|_|_| |_|\___/ \__|_|  \__,_|

*/

#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
  int maxDistToClosest(vector<int> &seats)
  {
    int n = seats.size();
    int maxDistance = 0;
    vector<int> left(n,INT_MAX), right(n, INT_MAX);

    left[0] = seats.front() ? 1 : 0;
    right[n-1] = seats.back() ? 1 : 0;

    for (int i = 1; i < n; i++)
    {
      int previousIdx = max(0, i-1);
      left[i] = seats[i] == 1 ? 0 : left[previousIdx] + 1;
    }
    
    for (int i = n-1; i >= 0; i--)
    {
      int nextIdx = min(n-1, i+1);
      right[i] = seats[i] == 1 ? 0 : right[nextIdx] + 1;
    }

    for(int i =0 ; i< n ; i++)
      maxDistance = max(maxDistance, min(left[i], right[i]));

    return maxDistance;
  }
};